### ISSUE TITLE
aria-pressed check too restrictive on toggle buttons

### ISSUE THREAD

**Issue by dd8** [validator/validator#1331]
The validator allows this code:

```
<input type="checkbox" role="button" aria-pressed="true" />
```

but not this:

```
<input type="checkbox" role="button" aria-pressed="mixed" />
<input type="checkbox" role="button" aria-pressed="false" />
```

**Issue:** `role=button` with `aria-pressed` indicates a toggle button, so how does the author indicate a toggled off state, or mixed state, if the only allowed state is `aria-pressed=true` ?

I think the basis for this check is [issue 1109](https://github.com/validator/validator/issues/1109), but I can't find the spec wording that requires aria-pressed=true:

> Finally, an input type=checkbox can have role=button, but only if it is paired with aria-pressed=true

**Comment by dd8** [comment]
I think it should allow `aria-pressed=true`, `aria-pressed=false`, `aria-pressed=mixed`

It's less clear what should happen when aria-pressed has no value or an empty value:

```
<input type=checkbox aria-pressed />
<input type=checkbox aria-pressed='' />
```

ARIA 1.2 says:

> Buttons with a non-empty [aria-pressed](https://www.w3.org/TR/wai-aria-1.2/#aria-pressed) attribute are toggle buttons

ARIA-in-HTML says this for input type=checkbox:

> Roles: [menuitemcheckbox](https://www.w3.org/TR/html-aria/#index-aria-menuitemcheckbox), [option](https://www.w3.org/TR/html-aria/#index-aria-option) or [switch](https://www.w3.org/TR/html-aria/#index-aria-switch); [button if used with aria-pressed](https://www.w3.org/TR/html-aria/#index-aria-button).

**Comment by scottaohara** [comment]
yes, the 3 allowed values is what this should check for, and are what should be allowed here.

if `aria-pressed` is undefined, e.g., `aria-pressed=""` or an invalid value, then ARIA states: "The element does not support being pressed."

In that case it would not be allowed as it wouldn't meet the intended requirement that if you're going to use a checkbox as a base to make a toggle button, then it _must_ be properly exposed as a toggle button.

**Comment by dd8** [comment]
This issue is still present - I'll investigate

**Comment by sideshowbarker** [comment]
Howdy @dd8 — have you had a chance to look back into this yet? Or you want me to take a look myself?

**Comment by dd8** [comment]
Ok, can see what the problem is - it's this code

https://github.com/validator/validator/blob/4657c6469a2e076ef1ad607c98f07c270841b470/src/nu/validator/checker/schematronequiv/Assertions.java#L3340

I think this can just be removed because aria.rnc enumerates the allowed values of aria-pressed:

https://github.com/validator/validator/blob/4657c6469a2e076ef1ad607c98f07c270841b470/schema/html5/aria.rnc#L340

**Comment by sideshowbarker** [comment]
>> I think it should allow `aria-pressed=true`, `aria-pressed=false`, `aria-pressed=mixed`
>
> yes, the 3 allowed values is what this should check for, and are what should be allowed here.

Do the spec requirements also allow it with `aria-pressed="undefined"`?

```html
<input type="checkbox" role="button" aria-pressed="undefined">
```

If not, can I have a pointer to whatever spec requirements state that’s now allowed?

**Comment by dd8** [comment]
My reading of the spec (ARIA 1.2 and current Editors Draft) is the allowed values for aria-pressed are: true/false/mixed/undefined

https://www.w3.org/TR/wai-aria-1.2/#aria-pressed
https://www.w3.org/TR/wai-aria-1.3/#aria-pressed
https://w3c.github.io/aria/#aria-pressed

This interpretation is consistent with what MDN says as well:
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-pressed

aria-checked allows the same tri-state values (true/false/mixed/undefined) and doesn't flag aria-checked='undefined' in the current VNU

```
<div role='checkbox' aria-checked='true' ></div>
<div role='checkbox' aria-checked='false' ></div>
<div role='checkbox' aria-checked='mixed' ></div>
<div role='checkbox' aria-checked='undefined' ></div>
```

**Comment by sideshowbarker** [comment]
Thanks. Yeah, while I wrote this: 

> If not, can I have a pointer to whatever spec requirements state that’s now allowed?

…there's a typo in there; I had _meant_ to write this: 

> If not, can I have a pointer to whatever spec requirements state that’s _not_ allowed?

So yeah, based on your response, it seems we're all good here already.
